compiler/src/dotty/tools/dotc/transform/init/Semantic.scala

```diff
@@ -0,0 +1,886 @@
+        eval(stats, thisV, klass).foreach { res => errorBuffer ++= res.errors }
+        errorBuffer ++= evalArgs(argss.flatten, thisV, klass)
+        superCall(tref, ctor, tree)
+
+      case tree @ NewExpr(tref, New(tpt), ctor, argss) =>
+        errorBuffer ++= evalArgs(argss.flatten, thisV, klass)
+        superCall(tref, ctor, tree)
+
+      case _ =>
+        val tref = typeRefOf(parent.tpe)
+        superCall(tref, tref.classSymbol.primaryConstructor, parent)
+    }
+
+    // see spec 5.1 about "Template Evaluation".
+    // https://www.scala-lang.org/files/archive/spec/2.13/05-classes-and-objects.html
+    if !klass.is(Flags.Trait) then
+      // 1. first init parent class recursively
+      // 2. initialize traits according to linearization order
+      val superParent = tpl.parents.head
+      val superCls = superParent.tpe.classSymbol.asClass
+      initParent(superParent)
+
+      val parents = tpl.parents.tail
+      val mixins = klass.baseClasses.tail.takeWhile(_ != superCls)
+      mixins.reverse.foreach { mixin =>
+        parents.find(_.tpe.classSymbol == mixin) match
+        case Some(parent) => initParent(parent)
+        case None =>
+          val tref = typeRefOf(klass.typeRef.baseType(mixin).typeConstructor)
+          val ctor = tref.classSymbol.primaryConstructor
+          if ctor.exists then superCall(tref, ctor, superParent)
+      }
+
+
+    // class body
+    tpl.body.foreach {
+      case vdef : ValDef if !vdef.symbol.is(Flags.Lazy) =>
+        val res = eval(vdef.rhs, thisV, klass, cacheResult = true)
+        errorBuffer ++= res.errors
+        thisV.updateField(vdef.symbol, res.value)
+
+      case _: MemberDef =>
+
+      case tree =>
+        errorBuffer ++= eval(tree, thisV, klass).errors
+    }
+
+    Result(thisV, errorBuffer.toList)
+  }
+
+  /** Check usage of terms inside types
+   *
+   *  This is intended to avoid type soundness issues in Dotty.
+   */
+  def checkTermUsage(tpt: Tree, thisV: Value, klass: ClassSymbol): Contextual[List[Error]] =
+    val buf = new mutable.ArrayBuffer[Error]
+    val traverser = new TypeTraverser {
+      def traverse(tp: Type): Unit = tp match {
+        case TermRef(_: SingletonType, _) =>
+          buf ++= cases(tp, thisV, klass, tpt).errors
+        case _ =>
+          traverseChildren(tp)
+      }
+    }
+    traverser.traverse(tpt.tpe)
+    buf.toList
+
+}
+
+object Semantic {
+
+// ----- Utility methods and extractors --------------------------------
+
+  def typeRefOf(tp: Type)(using Context): TypeRef = tp.dealias.typeConstructor match {
+    case tref: TypeRef => tref
+    case hklambda: HKTypeLambda => typeRefOf(hklambda.resType)
+  }
+
+  opaque type Arg  = Tree | ByNameArg
+  case class ByNameArg(tree: Tree)
+
+  extension (arg: Arg)
+    def isByName = arg.isInstanceOf[ByNameArg]
+    def tree: Tree = arg match
+      case t: Tree      => t
+      case ByNameArg(t) => t
+
+  object Call {
```

Comment:
Is there something like this extractor elsewhere in Dotty for general use? If not, should there be (should this be moved out of init where other code can use it)?